### PR TITLE
PR#58 (changeTitlePlaceholderText)

### PR DESCRIPTION
**Pull request by rgadon107**
## PR Summary

Adds unit and integration tests for the function `change_title_placeholder_text( $text, $post )` registered to the WordPress filter `'enter_title_here'` in the `tours` plugin ( relative path: `wp-content/plugins/tours/src/admin/edit-form-advanced.php` ).

### CODE REVIEW

File: plugins/tours/src/admin/edit-form-advanced.php [hunk 1]
```diff
@@ -11,22 +11,23 @@
...
  */
-function change_title_placeholder_text( $text ) {
-	if ( 'tours' !== get_post_type() ) {
+function change_title_placeholder_text( $text, $post ) {
+	if ( 'tours' !== get_post_type( $post ) ) {
 		return $text;
 	}
 
-	return '<em>' . 'Theme of this Cornerstone tour.' . '</em>';
+	return '<em>' . "Theme of this Cornerstone tour." . '</em>';
```
Comment (hellofromtonya) [inline plugins/tours/src/admin/edit-form-advanced.php:30]: There’s no need to concatenate as you have 3 string literals. Instead combine into one string. Why? Runs faster.

```php
return '<em>Theme of this Cornerstone tour.</em>';
```

File: plugins/tours/tests/phpunit/integration/admin/changeTitlePlaceholderText.php [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+ * @package spiralWebDb\CornerstoneTours\Tests\Integration
+ * @group   tours
+ * @group   admin
+ */
+class Tests_ChangeTitlePlaceholderText extends Test_Case {
+
+	/**
+	 * Set up each test.
+	 */
+	public function setUp() {
```
Comment (hellofromtonya) [inline plugins/tours/tests/phpunit/integration/admin/changeTitlePlaceholderText.php:28]: This is not needed as there’s no code in it.

File: plugins/tours/tests/phpunit/integration/admin/changeTitlePlaceholderText.php [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+		parent::setUp();
+	}
+
+	/**
+	 * Test change_title_placeholder_text() should return the given text when post type is 'post'.
+	 */
+	public function test_should_return_given_text_when_post_type_is_not_tours() {
+		// Create and get the post object via the factory method with 'post' post_type.
+		$post = self::factory()->post->create_and_get();
+		$expected = __( 'Add title' );
```
Comment (hellofromtonya) [inline plugins/tours/tests/phpunit/integration/admin/changeTitlePlaceholderText.php:38]: You can get rid of the i18n `__()`. Not needed here or in the filter.

File: plugins/tours/tests/phpunit/integration/admin/changeTitlePlaceholderText.php [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+		// Create and get the post object via the factory method and set post_type to 'tours' .
+		$post = $this->factory()->post->create_and_get( [ 'post_type' => 'tours' ] );
+		$expected_html = <<<PLACEHOLDER
+Theme of this Cornerstone tour.
+PLACEHOLDER;
+
+		// Run the output buffer to fire the event to which the callback is registered.
+		ob_start();
+		// Echo the value returned by the registered callback to compare in assert below.
+		echo apply_filters( 'enter_title_here', __( 'Add title' ), $post );
```
Comment (hellofromtonya) [inline plugins/tours/tests/phpunit/integration/admin/changeTitlePlaceholderText.php:62]: Same thing here.

File: plugins/tours/tests/phpunit/unit/admin/changeTitlePlaceholderText.php [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+	/**
+	 * Prepares the test environment before each test.
+	 */
+	protected function setUp() {
+		parent::setUp();
+
+		require_once TOURS_ROOT_DIR . '/src/admin/edit-form-advanced.php';
+
+		$this->post          = m::mock( 'post' );
```
Comment (hellofromtonya) [inline plugins/tours/tests/phpunit/unit/admin/changeTitlePlaceholderText.php:43]: Mockery mocks a class. `post` is not a class. Instead, it’s `m:::mock( WP_Post::class );`.

File: plugins/tours/tests/phpunit/unit/admin/changeTitlePlaceholderText.php [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+	/**
+	 * Test change_title_placeholder_text() should return modified text when post type is 'tours'.
+	 */
+	public function test_should_return_modified_text_when_post_type_is_tours() {
+		Functions\expect( 'get_post_type' )
+			->once()
+			->with( 'post' )
+			->andReturn( 'tours' );
+		$text     = 'Add title.';
+		$expected = '<em>' . "Theme of this Cornerstone tour." . '</em>';
```
Comment (hellofromtonya) [inline plugins/tours/tests/phpunit/unit/admin/changeTitlePlaceholderText.php:68]: Same thing here.

File: plugins/tours/src/admin/edit-form-advanced.php (hunk 1 above)
Comment (rgadon107) [reply plugins/tours/src/admin/edit-form-advanced.php:30]: @hellofromtonya Fixed.

File: plugins/tours/tests/phpunit/integration/admin/changeTitlePlaceholderText.php (hunk 2 above)
Comment (rgadon107) [reply plugins/tours/tests/phpunit/integration/admin/changeTitlePlaceholderText.php:28]: @hellofromtonya Fixed.

File: plugins/tours/tests/phpunit/integration/admin/changeTitlePlaceholderText.php (hunk 3 above)
Comment (rgadon107) [reply plugins/tours/tests/phpunit/integration/admin/changeTitlePlaceholderText.php:38]: @hellofromtonya Removed.

File: plugins/tours/tests/phpunit/integration/admin/changeTitlePlaceholderText.php (hunk 4 above)
Comment (rgadon107) [reply plugins/tours/tests/phpunit/integration/admin/changeTitlePlaceholderText.php:62]: @hellofromtonya Removed.

File: plugins/tours/tests/phpunit/unit/admin/changeTitlePlaceholderText.php (hunk 6 above)
Comment (rgadon107) [reply plugins/tours/tests/phpunit/unit/admin/changeTitlePlaceholderText.php:68]: @hellofromtonya Done.

File: plugins/tours/tests/phpunit/unit/admin/changeTitlePlaceholderText.php (hunk 5 above)
Comment (rgadon107) [reply plugins/tours/tests/phpunit/unit/admin/changeTitlePlaceholderText.php:43]: @hellofromtonya I agree with you that the proper expression of the class to be mocked is `WP_Post::class`. However, when I pass that as an argument into `Mockery::mock()`, the unit test fails. 

The filter docblock (screenshot below) lists the second parameter of the filter as the post object. 

wp-admin_edit-form-advanced

The PHP Manual says of the scope resolution operator ( double colon ): 

> The Scope Resolution Operator (also called Paamayim Nekudotayim) or in simpler terms, the double colon, is a token that allows access to static, constant, and overridden properties or methods of a class.

>When referencing these items from outside the class definition, use the name of the class.

>_As of PHP 5.3.0, it's possible to reference the class using a variable. The variable's value can not be a keyword (e.g. self, parent and static)._  (italics mine). 

The filter docblock uses the variable `$post` to represent `WP_Post::class` in the filter. When `post` is passed into Mockery as: 

>`$this->post  = m::mock( 'post' );`

the unit test passes. Unless you can come up with a workaround, I think we need to go with `$this->post  = m::mock( 'post' );`.

Comment (hellofromtonya) [reply plugins/tours/tests/phpunit/unit/admin/changeTitlePlaceholderText.php]: Mockery mocks an instance of a class, i.e. objects. `'post'` is not class name or an instance of the object. Why?

1. The variable `$post` you noted above is within WordPress Core. Your test does not have access to that variable. 
2. The variable within core is bound to an instance of which class? `WP_Post`. 
3. Therefore, you want to mock the class and not the variable.

[Check out the documentation](http://docs.mockery.io/en/latest/reference/creating_test_doubles.html#classes-abstracts-interfaces) for mocking classes, interfaces, and abstracts.

Note: For this particular test, you could also pass in `null` instead of mocking `WP_Post`. Why? Think about it. 

- The callback function is not type hinting what `$post` parameter data type is. Yes, it's identified in the `@param` DocBlock, but not within the function's parameter list. 
- The test mocks `get_post_type()` function. Therefore, `$post` can be anything. Why? It's not used within the unit test.

What do I mean "type hinting" `$post`?

This would be type hinting a function parameter:

```php
function change_title_placeholder_text( $text, WP_Post $post ) {
```

How does this work?
When the callback is invoked, PHP checks the data type of what's being passed to the 2nd parameter. Huh? It checks to make sure the data type for `$post` is `WP_Post`. If no, a fatal error is thrown. 

It's best practice to type hint your code. Why? Increases the quality of your code. How? Your code is more reliable and robust (quality attributes) because it makes sure that it receives the type of data it expects.

Comment (rgadon107) [reply plugins/tours/tests/phpunit/unit/admin/changeTitlePlaceholderText.php]: @hellofromtonya Thank you for responding back on this, and providing such a clear explanation.
